Examples/ReloadDataExample/ReloadDataViewController.swift

```diff
@@ -0,0 +1,151 @@
+//  CollectionKit
+//
+//  Created by yansong li on 2017-09-04.
+//  Copyright © 2017 lkzhao. All rights reserved.
+//
+
+import UIKit
+import CollectionKit
+
+let maxDataCount: Int = 1000
+
+class ReloadDataViewController: CollectionViewController {
+  var data: [Int] = []
+  var dataSet: Set<Int> = Set()
+
+  let addButton: UIButton = {
+    let button = UIButton()
+    let addAttributedTitle =
+        NSAttributedString(string: "+",
+                           attributes: [NSFontAttributeName : UIFont.boldSystemFont(ofSize: 20)])
+    button.setTitleColor(.black, for: .normal)
+    button.setAttributedTitle(addAttributedTitle, for: .normal)
+    button.backgroundColor = UIColor(hue: 0.16, saturation: 0.68, brightness: 0.98, alpha: 1)
+    return button
+  }()
+
+  let removeButton: UIButton = {
+    let button = UIButton()
+    let removeAttributedTitle =
+      NSAttributedString(string: "-",
+                         attributes: [NSFontAttributeName : UIFont.boldSystemFont(ofSize: 20)])
+    button.setAttributedTitle(removeAttributedTitle, for: .normal)
+    button.setTitleColor(.black, for: .normal)
+    button.backgroundColor = UIColor(hue: 0.20, saturation: 0.68, brightness: 0.98, alpha: 1)
+    return button
+  }()
+
+  let reloadButton: UIButton = {
+    let button = UIButton()
+    button.setTitle("Reset", for: .normal)
+    button.setTitleColor(.black, for: .normal)
+    button.backgroundColor = UIColor(hue: 0.22, saturation: 0.68, brightness: 0.98, alpha: 1)
+    return button
+  }()
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    view.backgroundColor = .white
+    addButton.addTarget(self, action: #selector(add), for: .touchUpInside)
+    view.addSubview(addButton)
+    removeButton.addTarget(self, action: #selector(remove), for: .touchUpInside)
+    view.addSubview(removeButton)
+    reloadButton.addTarget(self, action: #selector(reload), for: .touchUpInside)
+    view.addSubview(reloadButton)
+
+    let dataProvider = ArrayDataProvider(data: data, identifierMapper: { (_, data) in
+      return "\(data)"
+    })
+    let layout = FlowLayout<Int>(insets: UIEdgeInsetsMake(15, 15, 15, 15),
+                                 lineSpacing: 15,
+                                 interitemSpacing: 15)
+    let presenter = ZoomPresenter()
+    presenter.insertAnimation = .scale
+    presenter.deleteAnimation = .scale
+    provider = CollectionProvider(
+      dataProvider: dataProvider,
+      viewUpdater: { (view: UILabel, data: Int, index: Int) in
+        view.backgroundColor = UIColor(hue: CGFloat(data) / CGFloat(maxDataCount),
+                                       saturation: 0.68,
+                                       brightness: 0.98,
+                                       alpha: 1)
+        view.textColor = .white
+        view.textAlignment = .center
+        view.text = "\(data)"
+      },
+      layout: layout,
+      sizeProvider: { _, _, _ in
+        return CGSize(width: 60, height: 60)
```

Comment:
Let do some calculation here so they don't leave empty spaces on the right hand side.
A trick that I usually do is to divide the collection size with a preferred width to get a column count. Then use this column count to figure out the cell size.
sizeProvider: { _, _, collectionSize in 
  let columns = max(1, Int(collectionSize.width / 60))
  let cellWidth = collectionSize.width / columns
  return CGSize(width: cellWidth height: cellWidth)
}